examples/jsm/webxr/ARButton.js

```diff
@@ -1,161 +1,195 @@
-var ARButton = {
+class ARButton {
 
-	createButton: function ( renderer, sessionInit = {} ) {
+	constructor() {
 
-		function showStartAR( /*device*/ ) {
+		this.renderer = null;
+		this.sessionInit = null;
+		this.button = null;
+		this.message = null;
 
-			var currentSession = null;
+		this.currentSession = null;
 
-			function onSessionStarted( session ) {
+	}
 
-				session.addEventListener( 'end', onSessionEnded );
+	createButton( renderer, sessionInit = {} ) {
```

Comment:
I think it's better to just make `createButton()` to a static function. This code:
```js
const newARButton = new ARButton();
```
should not be necessary.

Reply:
I think I don't get it and I just tried your suggestion.
If we define createButton as a static method we lose the `this` reference and would be needed to `new VRButton` when importing the button and that will break any existing app which expects to get a full (somehow initialized in the constructor) instance and then calling its .createButton which has references to any this.func needed, even the listeners:
This change is fully backwards compatible.

Reply:
An app does not call `new VRButton`.  The intended usage is:
https://github.com/mrdoob/three.js/blob/29ec65b5223eb4a0cd155a6085e6ac7999b748d8/examples/webxr_vr_ballshooter.html#L90

This code seems to work just fine:
```js
class VRButton {

	static createButton( renderer, options ) {

		if ( options ) {

			console.error( 'THREE.VRButton: The "options" parameter has been removed. Please set the reference space type via renderer.xr.setReferenceSpaceType() instead.' );

		}

		function showEnterVR( /*device*/ ) {

			var currentSession = null;

			function onSessionStarted( session ) {

				session.addEventListener( 'end', onSessionEnded );

				renderer.xr.setSession( session );
				button.textContent = 'EXIT VR';

				currentSession = session;

			}

			function onSessionEnded( /*event*/ ) {

				currentSession.removeEventListener( 'end', onSessionEnded );

				button.textContent = 'ENTER VR';

				currentSession = null;

			}

			//

			button.style.display = '';

			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';

			button.textContent = 'ENTER VR';

			button.onmouseenter = function () {

				button.style.opacity = '1.0';

			};

			button.onmouseleave = function () {

				button.style.opacity = '0.5';

			};

			button.onclick = function () {

				if ( currentSession === null ) {

					// WebXR's requestReferenceSpace only works if the corresponding feature
					// was requested at session creation time. For simplicity, just ask for
					// the interesting ones as optional features, but be aware that the
					// requestReferenceSpace call will fail if it turns out to be unavailable.
					// ('local' is always available for immersive sessions and doesn't need to
					// be requested separately.)

					var sessionInit = { optionalFeatures: [ 'local-floor', 'bounded-floor', 'hand-tracking' ] };
					navigator.xr.requestSession( 'immersive-vr', sessionInit ).then( onSessionStarted );

				} else {

					currentSession.end();

				}

			};

		}

		function disableButton() {

			button.style.display = '';

			button.style.cursor = 'auto';
			button.style.left = 'calc(50% - 75px)';
			button.style.width = '150px';

			button.onmouseenter = null;
			button.onmouseleave = null;

			button.onclick = null;

		}

		function showWebXRNotFound() {

			disableButton();

			button.textContent = 'VR NOT SUPPORTED';

		}

		function stylizeElement( element ) {

			element.style.position = 'absolute';
			element.style.bottom = '20px';
			element.style.padding = '12px 6px';
			element.style.border = '1px solid #fff';
			element.style.borderRadius = '4px';
			element.style.background = 'rgba(0,0,0,0.1)';
			element.style.color = '#fff';
			element.style.font = 'normal 13px sans-serif';
			element.style.textAlign = 'center';
			element.style.opacity = '0.5';
			element.style.outline = 'none';
			element.style.zIndex = '999';

		}

		if ( 'xr' in navigator ) {

			var button = document.createElement( 'button' );
			button.id = 'VRButton';
			button.style.display = 'none';

			stylizeElement( button );

			navigator.xr.isSessionSupported( 'immersive-vr' ).then( function ( supported ) {

				supported ? showEnterVR() : showWebXRNotFound();

			} );

			return button;

		} else {

			var message = document.createElement( 'a' );

			if ( window.isSecureContext === false ) {

				message.href = document.location.href.replace( /^http:/, 'https:' );
				message.innerHTML = 'WEBXR NEEDS HTTPS'; // TODO Improve message

			} else {

				message.href = 'https://immersiveweb.dev/';
				message.innerHTML = 'WEBXR NOT AVAILABLE';

			}

			message.style.left = 'calc(50% - 90px)';
			message.style.width = '180px';
			message.style.textDecoration = 'none';

			stylizeElement( message );

			return message;

		}

	}

}

export { VRButton };

```

Reply:
Ahh I get what you mean, putting everything inside createButton, but thats not very ES6ish tbh...
What I wanted to create is a modern class as if we were creating it from scratch, while keeping it backward compatible as  having the fact that we need to live forever with the .createButton initialization approach.

Reply:
> but thats not very ES6ish tbh

Well, that's a matter of taste. Not everything has to be written as a class. 

TBH, I would keep the code as it is since I personally see no improvement in your approach. It's even worse than before since now a lot of methods are public. `createButton()` should be the only public interface that the button modules provide.

Reply:
> Not everything has to be written as a class

The fact this PR exists is because I thought you were just in the process of modernizing the codebase and transforming every component to a class, this approach is more ES6ish than just wrapping the same old fashioned function inside a dumb class container.

> Declaring non-public related stuff with an _ prefix is an acceptable tradeoff.

So this PR would be accepted defining non-createButton functions prefixed with underscore?

Reply:
Originally the class migration was started to improve tree shaking capabilities. But it turned out, this can be solved differently. I'm not sure anymore we want to go down the path and convert the entire codebase to classes if tree shaking can be improved differently.

> same old fashioned function 

Your argumentation seems a bit biased to me. Clean and modern code does not need to reside in a class. 

> So this PR would be accepted defining non-createButton functions prefixed with underscore?

That depends on @mrdoob preferences. I personally see no need for a change.

Reply:
@mrdoob In `WebGPURenderer` and its new components, I've decided to start with the class approach. However, if you prefer the existing style of `WebGLRenderer`, the code can be refactored. I know that not every dev is a fan of pseudo-private properties and methods achieved via the `_` prefix.

Reply:
> The fact this PR exists is because I thought you were just in the process of modernizing the codebase and transforming every component to a class

Yes, but we stopped when we realised we would be poluting the API with `_` variables because browsers don't support private properties in classes yet.

> @mrdoob In `WebGPURenderer` and its new components, I've decided to start with the class approach. However, if you prefer the existing style of `WebGLRenderer`, the code can be refactored.

Using `_` in `WebGPURenderer` is fine. Eventually we'll need to be converted them into `this.#`. Hopefully browsers that support webgpu will also support private properties in classes.

Reply:
Ok guys I get your point, I have modified the VRButton to follow the static way, if you like it i'll modify the ARButton too, otherwise feel free to just close the PR.

Anyway thanks for the amazing work on the library.

Reply:
`VRButton` is looking better. I think `() =>` are no longer needed though?